### PR TITLE
SI-9347  Efficient head/tail, if possible, for immutable maps & sets

### PR DESCRIPTION
**Pull request by Ichoran**
Most immutable collections, including sets and maps, have a better-than-O(n) method for removing an element.  In those cases, tail and possibly head were overridden so the head/tail pattern can be used with less of a performance penalty.

Speed improvements on head/tail pattern are (for sets/maps of size 1024, unless otherwise specified):

```
BitSet    190x
HashSet   250x
Set       400x
Set2        9x
Set4       12x
HashMap   430x
ListMap  2500x   // size 128
Map       430x
```

All tested sets/maps that were already fast are still fast.

Test code is reproduced below:

``` scala
object BenchTailSetMap {
  val th = new ichi.bench.Thyme

  val standard = 1 to 1024

  val sets = Map[String, Set[Int]](
    "Set" -> (Set.empty[Int] ++ standard),
    "Set4"-> Set(4, 7, 2, 1),
    "Set2"-> Set(3, 4),
    "HashSet" -> (collection.immutable.HashSet.empty[Int] ++ standard),
    "BitSet" -> (collection.immutable.BitSet.empty ++ standard),
    "SortedSet" -> (collection.immutable.SortedSet.empty[Int] ++ standard),
    "ListSet" -> (collection.immutable.ListSet.empty[Int] ++ standard)
  )

  val pairs = standard.map(i => i -> i.toString)

  // ListMap implementation is HORRIBLE, O(n^3) tail!  Cut down size.
  val maps = Map[String, Map[Int, String]](
    "Map" -> (Map.empty[Int, String] ++ pairs),
    "HashMap" -> (collection.immutable.HashMap.empty[Int, String] ++ pairs),
    "SortedMap" -> (collection.immutable.SortedMap.empty[Int, String] ++ pairs),
    "ListMap" -> (collection.immutable.ListMap.empty[Int, String] ++ pairs.take(128))
  )

  def hts(s: Set[Int]) = {
    var si = s
    var x = 0
    while (si.nonEmpty) {
      x += si.head
      si = si.tail
    }
    x
  }

  def htm(m: Map[Int, String]) = {
    var mi = m
    var x = 0
    while (mi.nonEmpty) {
      x += mi.head._2.length
      mi = mi.tail
    }
    x
  }

  def run() {
    sets.toList.sortBy(_._1).foreach{ case (name, s) => th.pbench(hts(s), s.size, name) }
    maps.toList.sortBy(_._1).foreach{ case (name, m) => th.pbench(htm(m), m.size, name) }
  }
}
```

### CODE REVIEW

**Comment by soc** [comment]
t7445:

```
java.lang.IllegalArgumentException: requirement failed
    at scala.Predef$.require(Predef.scala:212)
    at Test$.delayedEndpoint$Test$1(t7445.scala:5)
    at Test$delayedInit$body.apply(t7445.scala:3)
    at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
    at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
    at scala.App$class.scala$App$class$$$anonfun$1(App.scala:76)
    at scala.App$class.scala$App$class$$$anonfun$1$adapted(App.scala:76)
    at scala.App$class$$Lambda$132/1939990953.apply(Unknown Source)
    at scala.collection.immutable.List.foreach(List.scala:381)
    at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
    at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
    at scala.App$class.main(App.scala:76)
    at Test$.main(t7445.scala:3)
    at Test.main(t7445.scala)
```

t8549:

```
java.io.InvalidClassException: scala.collection.immutable.BitSet$BitSet1; local class incompatible: stream classdesc serialVersionUID = 2260107458435649300, local class serialVersionUID = 8615031960210779264
```

File: src/library/scala/collection/immutable/Set.scala [hunk 1]
```diff
@@ -191,6 +197,8 @@ object Set extends ImmutableSetFactory[Set] {
       else if (f(elem4)) Some(elem4)
       else None
     }
+    override def head: A = elem1
+    override def tail: Set[A] = new Set3(elem1, elem2, elem3)
```
Comment (janekdb) [inline src/library/scala/collection/immutable/Set.scala]: `elem2, elem3, elem4`

Comment (Ichoran) [inline src/library/scala/collection/immutable/Set.scala]: Ack, that's a bad one!  I'll fix it up as soon as I get a moment.

**Comment by Ichoran** [comment]
Hm, `ListSet` doesn't work the way I'd thought it did.  So the speedup doesn't materialize.  Maybe I should switch to a `Queue`-like strategy (which at least avoids O(n^2)).

**Comment by Ichoran** [comment]
@SethTisue - Maybe this one should come off hold if possible--it'd be nice to have the performance improvements in 2.12, I think?  (I resolved the ListMap-is-backwards issue by just documenting that if you want to take it apart quickly, you need to use last/init instead of head/tail.)

**Comment by SethTisue** [comment]
I had put it "on hold" to mark that the ball was in your court, not a reviewer's court. Or so I thought — what's the current status? Does 41d81ab resolve your earlier remark about "the speedup doesn't materialize"?

**Comment by Ichoran** [comment]
@SethTisue - Yes.  It "resolves" it by leaving head/tail slow and alerting people that last/init is how to be fast (and changing it so it actually is fast).

File: src/library/scala/collection/immutable/BitSet.scala [hunk 2]
```diff
@@ -103,6 +103,7 @@ object BitSet extends BitSetFactory[BitSet] {
     else new BitSetN(elems)
   }
 
+  @SerialVersionUID(2260107458435649300L)
```
Comment (SethTisue) [inline src/library/scala/collection/immutable/BitSet.scala:106]: does this have anything to do with the changes in the rest of the commit?

Comment (Ichoran) [inline src/library/scala/collection/immutable/BitSet.scala:106]: @SethTisue - Doubt it.  Not sure where that came from.  I'll check it out later when I have time.

**Comment by SethTisue** [comment]
just checking: this commit doesn't add any tests. did you look and see if there is already sufficient testing in place that changes like these are safe?

**Comment by Ichoran** [comment]
@SethTisue - Seemed like there were enough, but I'll run collections-laws on it to be (more) sure.

**Comment by Ichoran** [comment]
@SethTisue - I verified that the quasi-comprehensive collections tests still all pass.

File: src/library/scala/collection/BitSetLike.scala [hunk 3]
```diff
@@ -204,6 +204,27 @@ trait BitSetLike[+This <: BitSetLike[This] with SortedSet[Int]] extends SortedSe
...
+  override def last: Int = {
+    var i = nwords - 1
+    while (i >= 0) {
+      val wi = word(i)
+      if (wi != 0L) return WordLength*i + 63 - java.lang.Long.numberOfLeadingZeros(wi)
+      i += 1
+    }
+    throw new NoSuchElementException("Empty BitSet")
+  }
+
```
Comment (plokhotnyuk) [inline src/library/scala/collection/BitSetLike.scala:227]: I would replace it by tail recursion to avoid `var` & `return` w/ zero overhead: 

``` scala
override def head: Int = head(nwords, 0)

@tailrec private def head(n: Int, i: Int): Int =
  if (i < n) {
   val wi = word(i)
   if (wi != 0L) WordLength*i + java.lang.Long.numberOfTrailingZeros(wi)
   else head(n, i + 1)
  } else throw new NoSuchElementException("Empty BitSet")
```

Comment (Ichoran) [inline src/library/scala/collection/BitSetLike.scala:227]: The implementation is longer and harder to understand due to having to split a single method into two.  (Also, note that there is no reason to pass `n` as a parameter instead of just using `nwords` directly, as the logic for `last` is different and cannot share the same method.)

I don't think this is a win in this case.

File: src/library/scala/collection/BitSetLike.scala [hunk 4]
```diff
@@ -204,6 +204,27 @@ trait BitSetLike[+This <: BitSetLike[This] with SortedSet[Int]] extends SortedSe
   def subsetOf(other: BitSet): Boolean =
     (0 until nwords) forall (idx => (this.word(idx) & ~ other.word(idx)) == 0L)
 
+  override def head: Int = {
```
Comment (szeiger) [inline src/library/scala/collection/BitSetLike.scala:207]: The implementation of this algorithm is slightly different from what `BitSetLike.foreach` does. Should they be the same? Which one is faster?

Comment (Ichoran) [inline src/library/scala/collection/BitSetLike.scala:207]: Generally, scanning bit by bit is slower than using `numberOfTrailingZeros` unless the bits are pretty dense.  It might be worth investigating whether `foreach` could be sped up, and if so, to make a separate commit for it.

**Comment by szeiger** [comment]
LGTM
